File: examples/php_arbitary_project/acme_billing.php
<?php

declare(strict_types=1);

require __DIR__ . '/vendor/autoload.php';

use Fahara02\UdbLaravel\UdbClient;
use Fahara02\UdbLaravel\UdbMetadata;
use Google\Protobuf\Struct;
use PhpArbitaryProject\Acme\Billing\V1\Product;
use Udb\Entity\V1\Chunk;
use Udb\Entity\V1\DeleteRequest;
use Udb\Entity\V1\ObjectRequest;
use Udb\Entity\V1\SelectRequest;
use Udb\Entity\V1\UpsertRequest;
use Udb\Entity\V1\UrlRequest;
use Udb\Entity\V1\VectorPointMutation;
use Udb\Entity\V1\VectorSearchRequest;
use Udb\Entity\V1\VectorUpsertRequest;

$target = getenv('UDB_TARGET') ?: '127.0.0.1:50051';
$projectId = 'default';
$tenantId = 'acme-org-1';
$messageType = 'acme.billing.v1.Product';
$collection = 'acme_products';
$bucket = 'acme-billing-documents';
$showTimings = filter_var(getenv('UDB_SHOW_TIMINGS') ?: 'false', FILTER_VALIDATE_BOOLEAN);
$warmup = filter_var(getenv('UDB_WARMUP') ?: 'true', FILTER_VALIDATE_BOOLEAN);

$client = new UdbClient([
    'endpoint' => $target,
    'tls' => ['enabled' => false],
    'deadline_ms' => 30_000,
]);

$metadata = new UdbMetadata(
    tenantId: $tenantId,
    userId: 'sdk-php-example',
    purpose: 'billing.example',
    correlationId: 'php-acme-billing-example',
    scopes: [
        'udb:read',
        'udb:write',
        'udb:admin',
        'udb:vector:read',
        'udb:vector:write',
        'udb:object:presign',
        'udb:stream',
    ],
    serviceIdentity: 'examples.php',
    projectId: $projectId,
    clientCatalogVersion: '1.0.0',
);

if ($warmup) {
    timed('client warmup', fn() => $client->warmup($metadata));
}

$runId = (string) time() . random_int(1000, 9999);
$productId = "prod-sdk-php-crud-{$runId}";

timed('relational cleanup delete', function () use ($client, $metadata, $messageType, $productId, $runId): void {
    $client->delete(new DeleteRequest([
    'message_type' => $messageType,
    'filter' => productFilter($productId),
    'idempotency_key' => "php-product-cleanup-{$runId}",
    ]), $metadata);
});

$product = new Product([
    'product_id' => $productId,
    'name' => 'SDK smoke test product',
    'description' => 'Inserted by the PHP UDB SDK example',
    'price_cents' => 22900,
    'sku' => 'SDK-PHP-001',
]);

$upsert = new UpsertRequest([
    'message_type' => $messageType,
    'record_json' => productRecordJson($product),
    'conflict_fields' => ['product_id'],
    'return_record' => true,
    'idempotency_key' => "php-product-create-{$runId}",
]);

timed('relational create upsert', fn() => $client->upsert($upsert, $metadata));

$rows = timed('relational create select', fn() => $client->select(new SelectRequest([
    'message_type' => $messageType,
    'filter' => productFilter($productId),
    'limit' => 10,
]), $metadata));

$rowCount = recordCount($rows);
if ($rowCount !== 1) {
    throw new RuntimeException("relational create/read expected 1 row, got {$rowCount}");
}
printf("relational create/read rows=%d\n", $rowCount);

$product->setName('SDK smoke test product updated');
timed('relational update upsert', fn() => $client->upsert(new UpsertRequest([
    'message_type' => $messageType,
    'record_json' => productRecordJson($product),
    'conflict_fields' => ['product_id'],
    'return_record' => true,
    'idempotency_key' => "php-product-update-{$runId}",
]), $metadata));

$rows = timed('relational update select', fn() => $client->select(new SelectRequest([
    'message_type' => $messageType,
    'filter' => productFilter($productId),
    'limit' => 10,
]), $metadata));
if (! recordsContain($rows, 'SDK smoke test product updated')) {
    throw new RuntimeException('relational update was not visible in select response');
}
printf("relational update verified rows=%d\n", recordCount($rows));

timed('relational delete', fn() => $client->delete(new DeleteRequest([
    'message_type' => $messageType,
    'filter' => productFilter($productId),
    'idempotency_key' => "php-product-delete-{$runId}",
]), $metadata));

$rows = timed('relational delete select', fn() => $client->select(new SelectRequest([
    'message_type' => $messageType,
    'filter' => productFilter($productId),
    'limit' => 10,
]), $metadata));
$rowCount = recordCount($rows);
if ($rowCount !== 0) {
    throw new RuntimeException("relational delete expected 0 rows, got {$rowCount}");
}
printf("relational delete verified rows=%d\n", $rowCount);

$vector = makeVector(768);
timed('vector upsert', fn() => waitUnary($client->stub()->VectorUpsert(new VectorUpsertRequest([
    'collection' => $collection,
    'points' => [
        new VectorPointMutation([
            'id' => '22222222-2222-4222-8222-222222222222',
            'vector' => $vector,
        ]),
    ],
    'idempotency_key' => 'php-vector-sdk-001',
]), $metadata->toGrpcMetadata()), 'VectorUpsert'));

$vectorRows = timed('vector search', fn() => waitUnary($client->stub()->VectorSearch(new VectorSearchRequest([
    'collection' => $collection,
    'vector' => $vector,
    'limit' => 3,
    'with_payload' => true,
]), $metadata->toGrpcMetadata()), 'VectorSearch'));

printf("vector search points=%d\n", count($vectorRows->getPoints()));

$objectCall = $client->stub()->PutObject($metadata->toGrpcMetadata());
$objectCall->write(new Chunk([
    'bucket' => $bucket,
    'object_key' => 'invoices/sdk/php/smoke.txt',
    'data' => "hello from the PHP UDB SDK example\n",
    'final_chunk' => true,
    'content_type' => 'text/plain',
    'idempotency_key' => 'php-object-sdk-001',
]));
timed('object put', fn() => waitUnary($objectCall, 'PutObject'));

$objectBytes = timed('object get', function () use ($client, $metadata, $bucket): string {
    $objectStream = $client->stub()->GetObject(new ObjectRequest([
        'bucket' => $bucket,
        'object_key' => 'invoices/sdk/php/smoke.txt',
    ]), $metadata->toGrpcMetadata());
    $bytes = '';
    foreach ($objectStream->responses() as $chunk) {
        $bytes .= $chunk->getData();
    }
    assertOkStatus($objectStream->getStatus(), 'GetObject');
    return $bytes;
});
if (! str_contains($objectBytes, 'hello from the PHP UDB SDK example')) {
    throw new RuntimeException('object readback did not match uploaded content');
}
printf("object readback bytes=%d\n", strlen($objectBytes));

$url = timed('object presign', fn() => waitUnary($client->stub()->GeneratePresignedUrl(new UrlRequest([
    'bucket' => $bucket,
    'object_key' => 'invoices/sdk/php/smoke.txt',
    'method' => 'GET',
    'ttl_seconds' => 300,
    'content_type' => 'text/plain',
]), $metadata->toGrpcMetadata()), 'GeneratePresignedUrl'));

printf("object presigned url expires_at=%d\n", $url->getExpiresAtUnix());

/**
 * @return list<float>
 */
function makeVector(int $dimension): array
{
    $vector = [];
    for ($i = 0; $i < $dimension; $i++) {
        $vector[] = (($i % 17) + 1) / 17.0;
    }
    return $vector;
}

function productRecordJson(Product $product): string
{
    return json_encode([
        'product_id' => $product->getProductId(),
        'name' => $product->getName(),
        'description' => $product->getDescription(),
        'price_cents' => $product->getPriceCents(),
        'sku' => $product->getSku(),
    ], JSON_THROW_ON_ERROR);
}

function productFilter(string $productId): Struct
{
    $filter = new Struct();
    $filter->mergeFromJsonString(json_encode([
        'product_id' => $productId,
    ], JSON_THROW_ON_ERROR));
    return $filter;
}

function recordCount(object $rows): int
{
    $jsonRows = count($rows->getRecordsJson());
    if ($jsonRows > 0) {
        return $jsonRows;
    }
    return count($rows->getRows());
}

function recordsContain(object $rows, string $needle): bool
{
    foreach ($rows->getRecordsJson() as $record) {
        if (str_contains((string) $record, $needle)) {
            return true;
        }
    }
    foreach ($rows->getRows() as $row) {
        if (str_contains($row->serializeToJsonString(), $needle)) {
            return true;
        }
    }
    return false;
}

function waitUnary($call, string $rpcName): object
{
    [$response, $status] = $call->wait();
    assertOkStatus($status, $rpcName);
    if (! is_object($response)) {
        throw new RuntimeException("{$rpcName} returned no response");
    }
    return $response;
}

function assertOkStatus(object|array $status, string $rpcName): void
{
    $code = is_object($status) ? ($status->code ?? -1) : ($status['code'] ?? -1);
    if ($code === 0) {
        return;
    }
    $details = is_object($status) ? ($status->details ?? 'unknown gRPC error') : ($status['details'] ?? 'unknown gRPC error');
    throw new RuntimeException("{$rpcName} failed: {$details}");
}

/**
 * @template T
 * @param callable(): T $operation
 * @return T
 */
function timed(string $label, callable $operation): mixed
{
    global $showTimings;

    if (! $showTimings) {
        return $operation();
    }

    $start = hrtime(true);
    try {
        return $operation();
    } finally {
        $elapsedMs = (hrtime(true) - $start) / 1_000_000;
        printf("timing %-28s %.2f ms\n", $label, $elapsedMs);
    }
}

File: examples/python_arbitary_project/acme_billing.py
from __future__ import annotations

import os
import time
from typing import Any

from google.protobuf.json_format import MessageToJson

from gen.acme.billing.v1 import acme_billing_v1_pb2 as acme_billing
from udb.entity.v1 import types_pb2
from udb_client import Metadata, UdbClient, decode_records, to_struct
from udb_client.models import MetadataModel

TARGET_DEFAULT = "127.0.0.1:50051"
PROJECT_ID = "default"
MESSAGE_TYPE = "acme.billing.v1.Product"
COLLECTION = "acme_products"
BUCKET = "acme-billing-documents"


def main() -> None:
    target = os.getenv("UDB_TARGET", TARGET_DEFAULT)
    show_timings = os.getenv("UDB_SHOW_TIMINGS", "false").lower() == "true"
    warmup = os.getenv("UDB_WARMUP", "true").lower() == "true"

    metadata = MetadataModel(
        tenant_id="acme-org-1",
        user_id="sdk-python-example",
        purpose="billing.example",
        correlation_id="python-acme-billing-example",
        scopes=(
            "udb:read",
            "udb:write",
            "udb:admin",
            "udb:vector:read",
            "udb:vector:write",
            "udb:object:presign",
            "udb:stream",
        ),
        service_identity="examples.python",
        project_id=PROJECT_ID,
        client_catalog_version="1.0.0",
    ).to_metadata()

    with UdbClient(target, metadata) as client:
        if warmup:
            timed("client warmup", lambda: client.warmup(), show_timings)

        run_id = time.time_ns()
        product_id = f"prod-sdk-python-crud-{run_id}"

        timed(
            "relational cleanup delete",
            lambda: client.delete(
                message_type=MESSAGE_TYPE,
                filter={"product_id": product_id},
                idempotency_key=f"python-product-cleanup-{run_id}",
            ),
            show_timings,
        )

        product = acme_billing.Product(
            product_id=product_id,
            name="SDK smoke test product",
            description="Inserted by the Python UDB SDK example",
            price_cents=32900,
            sku="SDK-PY-001",
        )
        timed(
            "relational create upsert",
            lambda: client.upsert(
                message_type=MESSAGE_TYPE,
                record=product_json(product),
                conflict_fields=("product_id",),
                return_record=True,
                idempotency_key=f"python-product-create-{run_id}",
            ),
            show_timings,
        )

        rows = timed(
            "relational create select",
            lambda: client.select(
                message_type=MESSAGE_TYPE,
                filter={"product_id": product_id},
                limit=10,
            ),
            show_timings,
        )
        if record_count(rows) != 1:
            raise RuntimeError(f"relational create/read expected 1 row, got {record_count(rows)}")
        print(f"relational create/read rows={record_count(rows)}")

        product.name = "SDK smoke test product updated"
        timed(
            "relational update upsert",
            lambda: client.upsert(
                message_type=MESSAGE_TYPE,
                record=product_json(product),
                conflict_fields=("product_id",),
                return_record=True,
                idempotency_key=f"python-product-update-{run_id}",
            ),
            show_timings,
        )
        rows = timed(
            "relational update select",
            lambda: client.select(
                message_type=MESSAGE_TYPE,
                filter={"product_id": product_id},
                limit=10,
            ),
            show_timings,
        )
        if not records_contain(rows, "SDK smoke test product updated"):
            raise RuntimeError("relational update was not visible in select response")
        print(f"relational update verified rows={record_count(rows)}")

        timed(
            "relational delete",
            lambda: client.delete(
                message_type=MESSAGE_TYPE,
                filter={"product_id": product_id},
                idempotency_key=f"python-product-delete-{run_id}",
            ),
            show_timings,
        )
        rows = timed(
            "relational delete select",
            lambda: client.select(
                message_type=MESSAGE_TYPE,
                filter={"product_id": product_id},
                limit=10,
            ),
            show_timings,
        )
        if record_count(rows) != 0:
            raise RuntimeError(f"relational delete expected 0 rows, got {record_count(rows)}")
        print(f"relational delete verified rows={record_count(rows)}")

        vector = make_vector(768)
        timed(
            "vector upsert",
            lambda: client.vector_upsert(
                COLLECTION,
                [
                    types_pb2.VectorPointMutation(
                        id="33333333-3333-4333-8333-333333333333",
                        vector=vector,
                        payload=to_struct(
                            {
                                "product_id": "prod-sdk-python-001",
                                "name": "Python SDK vector point",
                            }
                        ),
                    )
                ],
                idempotency_key="python-vector-sdk-001",
            ),
            show_timings,
        )
        vector_rows = timed(
            "vector search",
            lambda: client.vector_search(COLLECTION, vector, limit=3, with_payload=True),
            show_timings,
        )
        print(f"vector search points={len(vector_rows.points)}")

        timed(
            "object put",
            lambda: client.put_object(
                BUCKET,
                "invoices/sdk/python/smoke.txt",
                b"hello from the Python UDB SDK example\n",
                content_type="text/plain",
                idempotency_key="python-object-sdk-001",
            ),
            show_timings,
        )
        object_bytes = timed(
            "object get",
            lambda: client.get_object(BUCKET, "invoices/sdk/python/smoke.txt"),
            show_timings,
        )
        if b"hello from the Python UDB SDK example" not in object_bytes:
            raise RuntimeError("object readback did not match uploaded content")
        print(f"object readback bytes={len(object_bytes)}")

        url = timed(
            "object presign",
            lambda: client.generate_presigned_url(
                BUCKET,
                "invoices/sdk/python/smoke.txt",
                method="GET",
                ttl_seconds=300,
                content_type="text/plain",
            ),
            show_timings,
        )
        print(f"object presigned url expires_at={url.expires_at_unix}")


def product_json(product: acme_billing.Product) -> bytes:
    return MessageToJson(
        product,
        preserving_proto_field_name=True,
        sort_keys=True,
        indent=None,
    ).encode("utf-8")


def record_count(rows: types_pb2.RecordSet) -> int:
    if rows.records_json:
        return len(rows.records_json)
    return len(rows.rows)


def records_contain(rows: types_pb2.RecordSet, needle: str) -> bool:
    if any(needle in record.decode("utf-8", errors="replace") for record in rows.records_json):
        return True
    return any(needle in str(row) for row in rows.rows)


def make_vector(dimension: int) -> list[float]:
    return [float((index % 17) + 1) / 17.0 for index in range(dimension)]


def timed(label: str, operation: Any, show_timings: bool) -> Any:
    if not show_timings:
        return operation()
    start = time.perf_counter()
    try:
        return operation()
    finally:
        elapsed_ms = (time.perf_counter() - start) * 1000.0
        print(f"timing {label:<28} {elapsed_ms:.2f} ms")


if __name__ == "__main__":
    main()
